README.md
# AngularJS ui-select [![Build Status](https://travis-ci.org/granteagon/ui-select-temp.svg?branch=master)](https://travis-ci.org/granteagon/ui-select-temp) [![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/granteagon/ui-select-temp?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

AngularJS-native version of [Select2](http://ivaynberg.github.io/select2/) and [Selectize](http://brianreavis.github.io/selectize.js/). [http://angular-ui.github.io/ui-select/](http://angular-ui.github.io/ui-select/)

[Getting Started](https://github.com/granteagon/ui-select-temp/wiki/Getting-Started)

- [Examples](http://angular-ui.github.io/ui-select/#examples)
- [Examples Source](./docs/examples)
- [Documentation](https://github.com/granteagon/ui-select-temp/wiki)

## Latest Changes

- Check [CHANGELOG.md](/CHANGELOG.md)

## Features

- Search, Select, Multi-select and Tagging
- Multiple Themes: Bootstrap, Select2 and Selectize
- Keyboard support
- No jQuery required (except for old browsers)
- Small code base: 4.57KB min/gzipped vs 20KB for select2

For the roadmap, check [issue #3](https://github.com/granteagon/ui-select-temp/issues/3) and the [Wiki page](https://github.com/granteagon/ui-select-temp/wiki/Roadmap).

## Installation Methods

### npm
```
$ npm install ui-select
```
### bower
```
$ bower install angular-ui-select
```

## Development

### Prepare your environment
* Install [Node.js](http://nodejs.org/) and NPM (should come with)
* Install global dev dependencies: `npm install -g gulp`
* Install local dev dependencies: `npm install` in repository directory

### Development Commands

* `gulp` to jshint, build and test
* `gulp build` to jshint and build
* `gulp test` for one-time test with karma (also build and jshint)
* `gulp watch` to watch src files to jshint, build and test when changed
* `gulp docs` build docs and examples

## Contributing

- Check [CONTRIBUTING.md](/CONTRIBUTING.md)
- Run the tests
- Try the [examples](./docs/examples)

When issuing a pull request, please exclude changes from the "dist" folder to avoid merge conflicts.


CHANGELOG.md
<a name="0.19.7"></a>
## [0.19.7](https://github.com/angular-ui/ui-select/compare/0.19.7...v0.19.7) (2017-02-03)




<a name="0.19.5"></a>
## [0.19.5](https://github.com/granteagon/ui-select-temp/compare/v0.19.5...v0.19.5) (2016-10-24)

## Reverted

* **Bug fix:** Search input isn't blocked ([0d81493](https://github.com/granteagon/ui-select-temp/commit/0d81493))

<a name="0.19.4"></a>
## [0.19.4](https://github.com/granteagon/ui-select-temp/compare/v0.19.4...v0.19.4) (2016-10-19)

### Bug Fixes

* **bootstrap:** add aria-expanded to the multiple select ([6766754](https://github.com/granteagon/ui-select-temp/commit/6766754)), closes [#1775](https://github.com/granteagon/ui-select-temp/issues/1775)
* ensure aria-activedescendant is correct ([e3be4d1](https://github.com/granteagon/ui-select-temp/commit/e3be4d1))
* only apply listbox role when open ([e902ffc](https://github.com/granteagon/ui-select-temp/commit/e902ffc))
* **bootstrap:** add search role ([f3194bf](https://github.com/granteagon/ui-select-temp/commit/f3194bf))

### Features

* **selectize:** add support for multiple selection  ([ff8071d](https://github.com/granteagon/ui-select-temp/commit/ff8071d)), closes [#295](https://github.com/granteagon/ui-select-temp/issues/295) [#1787](https://github.com/granteagon/ui-select-temp/issues/1787)

<a name="0.19.3"></a>
## [0.19.3](https://github.com/granteagon/ui-select-temp/compare/v0.19.3...v0.19.3) (2016-08-17)


<a name="0.19.2"></a>
## [0.19.2](https://github.com/granteagon/ui-select-temp/compare/v0.19.0...v0.19.2) (2016-08-16)


### Bug Fixes

* **bootstrap layout:** Restrict selected length to ui-select-container (#1680) ([01055c5](https://github.com/granteagon/ui-select-temp/commit/01055c5)), closes [#1576](https://github.com/granteagon/ui-select-temp/issues/1576)
* **release:** bump version with built files ([76cf9c3](https://github.com/granteagon/ui-select-temp/commit/76cf9c3))

<a name="0.19.1"></a>
## [0.19.1](https://github.com/granteagon/ui-select-temp/compare/v0.19.0...v0.19.1) (2016-08-09)

### Bug Fixes

* **bootstrap layout:** Restrict selected length to ui-select-container (#1680) ([01055c5](https://github.com/granteagon/ui-select-temp/commit/01055c5)), closes [#1576](https://github.com/granteagon/ui-select-temp/issues/1576)

<a name="0.19.0"></a>
# [0.19.0](https://github.com/granteagon/ui-select-temp/compare/v0.19.0...v0.19.0) (2016-08-07)

### Bug Fixes

* **bootstrap layout:** Restrict selected length to ui-select-container (#1680) ([01055c5](https://github.com/granteagon/ui-select-temp/commit/01055c5)), closes [#1576](https://github.com/granteagon/ui-select-temp/issues/1576)

<a name="0.18.1"></a>
## [0.18.1](https://github.com/granteagon/ui-select-temp/compare/v0.18.1...v0.18.1) (2016-08-07)

### Bug Fixes

* **bootstrap:** remove anchor tag in choices ([b15189d](https://github.com/granteagon/ui-select-temp/commit/b15189d))
* **uiSelect:** remove flicker on search change ([81c33d0](https://github.com/granteagon/ui-select-temp/commit/81c33d0)), closes [#1298](https://github.com/granteagon/ui-select-temp/issues/1298) [#1594](https://github.com/granteagon/ui-select-temp/issues/1594) [#1557](https://github.com/granteagon/ui-select-temp/issues/1557)
* **uiSelectController:** Select by click on non-multiple tagging (bis) (#1727) ([3dfde71](https://github.com/granteagon/ui-select-temp/commit/3dfde71))

### Features

* **events:** add open-close callback  ([21bcd5e](https://github.com/granteagon/ui-select-temp/commit/21bcd5e)), closes [#432](https://github.com/granteagon/ui-select-temp/issues/432) [#1153](https://github.com/granteagon/ui-select-temp/issues/1153)

### Performance Improvements

* **repeatParserService:** track groups by name  ([1770038](https://github.com/granteagon/ui-select-temp/commit/1770038)), closes [#1721](https://github.com/granteagon/ui-select-temp/issues/1721) [#1722](https://github.com/granteagon/ui-select-temp/issues/1722)

<a name="0.18.1"></a>
## [0.18.1](https://github.com/granteagon/ui-select-temp/compare/v0.18.0...v0.18.1) (2016-07-10)

### Bug Fixes

* **isDisabled:** do not modify item ([b95bf9f](https://github.com/granteagon/ui-select-temp/commit/b95bf9f)), closes [#1200](https://github.com/granteagon/ui-select-temp/issues/1200) [#1661](https://github.com/granteagon/ui-select-temp/issues/1661)
* **isLocked:** do not modify item ([c01d363](https://github.com/granteagon/ui-select-temp/commit/c01d363)), closes [#1269](https://github.com/granteagon/ui-select-temp/issues/1269) [#514](https://github.com/granteagon/ui-select-temp/issues/514)
* **removeSelected:** fix incorrect removal of preselected item  ([32b7924](https://github.com/granteagon/ui-select-temp/commit/32b7924)), closes [#1672](https://github.com/granteagon/ui-select-temp/issues/1672)
* **searchEnabled:** watch evaluated attribute value  ([4503295](https://github.com/granteagon/ui-select-temp/commit/4503295)), closes [#505](https://github.com/granteagon/ui-select-temp/issues/505)
* **select2:** Up-direction when using global theme  ([5336dc5](https://github.com/granteagon/ui-select-temp/commit/5336dc5)), closes [#1674](https://github.com/granteagon/ui-select-temp/issues/1674)
* **Selectize:** hide input box when selected data is 0 ([e179dc6](https://github.com/granteagon/ui-select-temp/commit/e179dc6)), closes [#1304](https://github.com/granteagon/ui-select-temp/issues/1304)
* **tagging:** infite digest loops when name is similar ([fcd9bc5](https://github.com/granteagon/ui-select-temp/commit/fcd9bc5)), closes [#1693](https://github.com/granteagon/ui-select-temp/issues/1693)
* **tagging:** Support paste with tagging enabled and tagging-label="false" ([668a0f3](https://github.com/granteagon/ui-select-temp/commit/668a0f3)), closes [#1668](https://github.com/granteagon/ui-select-temp/issues/1668)
* **uiSelectMultiple:** $select.refreshItems is not a function  ([a41a7fc](https://github.com/granteagon/ui-select-temp/commit/a41a7fc))

karma.conf.js
module.exports = function(config) {
  config.set({

    // Base path, that will be used to resolve files and exclude
    basePath: '',

    // Frameworks to use
    frameworks: ['jasmine'],

    // List of files / patterns to load in the browser
    files: [
      'node_modules/jquery/dist/jquery.js',
      'node_modules/angular/angular.js',
      'node_modules/angular-sanitize/angular-sanitize.js',
      'node_modules/angular-mocks/angular-mocks.js',

      'dist/select.js',
      'test/helpers.js',
      'test/**/*.spec.js'
    ],

    // List of files to exclude
    exclude: ['./index.js'],

    // Web server port
    port: 9876,

    // Level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,

    // Enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,

    // Start these browsers, currently available:
    // - Chrome
    // - ChromeCanary
    // - Firefox
    // - Opera
    // - Safari (only Mac)
    // - PhantomJS
    // - IE (only Windows)
    browsers: [process.env.TRAVIS ? 'Firefox' : 'Chrome'],
    reporters: ['spec'],

    // Continuous Integration mode
    // if true, it capture browsers, run tests and exit
    singleRun: false
  });
};


src/uisRepeatParserService.js
/**
 * Parses "repeat" attribute.
 *
 * Taken from AngularJS ngRepeat source code
 * See https://github.com/angular/angular.js/blob/v1.2.15/src/ng/directive/ngRepeat.js#L211
 *
 * Original discussion about parsing "repeat" attribute instead of fully relying on ng-repeat:
 * https://github.com/granteagon/ui-select-temp/commit/5dd63ad#commitcomment-5504697
 */

uis.service('uisRepeatParser', ['uiSelectMinErr','$parse', function(uiSelectMinErr, $parse) {
  var self = this;

  /**
   * Example:
   * expression = "address in addresses | filter: {street: $select.search} track by $index"
   * itemName = "address",
   * source = "addresses | filter: {street: $select.search}",
   * trackByExp = "$index",
   */
  self.parse = function(expression) {


    var match;
    //var isObjectCollection = /\(\s*([\$\w][\$\w]*)\s*,\s*([\$\w][\$\w]*)\s*\)/.test(expression);
    // If an array is used as collection

    // if (isObjectCollection){
    // 000000000000000000000000000000111111111000000000000000222222222222220033333333333333333333330000444444444444444444000000000000000055555555555000000000000000000000066666666600000000
    match = expression.match(/^\s*(?:([\s\S]+?)\s+as\s+)?(?:([\$\w][\$\w]*)|(?:\(\s*([\$\w][\$\w]*)\s*,\s*([\$\w][\$\w]*)\s*\)))\s+in\s+(\s*[\s\S]+?)?(?:\s+track\s+by\s+([\s\S]+?))?\s*$/);

    // 1 Alias
    // 2 Item
    // 3 Key on (key,value)
    // 4 Value on (key,value)
    // 5 Source expression (including filters)
    // 6 Track by

    if (!match) {
      throw uiSelectMinErr('iexp', "Expected expression in form of '_item_ in _collection_[ track by _id_]' but got '{0}'.",
              expression);
    }
    
    var source = match[5], 
        filters = '';

    // When using (key,value) ui-select requires filters to be extracted, since the object
    // is converted to an array for $select.items 
    // (in which case the filters need to be reapplied)
    if (match[3]) {
      // Remove any enclosing parenthesis
      source = match[5].replace(/(^\()|(\)$)/g, '');
      // match all after | but not after ||
      var filterMatch = match[5].match(/^\s*(?:[\s\S]+?)(?:[^\|]|\|\|)+([\s\S]*)\s*$/);
      if(filterMatch && filterMatch[1].trim()) {
        filters = filterMatch[1];
        source = source.replace(filters, '');
      }      
    }

    return {
      itemName: match[4] || match[2], // (lhs) Left-hand side,
      keyName: match[3], //for (key, value) syntax
      source: $parse(source),
      filters: filters,
      trackByExp: match[6],
      modelMapper: $parse(match[1] || match[4] || match[2]),
      repeatExpression: function (grouped) {
        var expression = this.itemName + ' in ' + (grouped ? '$group.items' : '$select.items');
        if (this.trackByExp) {
          expression += ' track by ' + this.trackByExp;
        }
        return expression;
      } 
    };

  };

  self.getGroupNgRepeatExpression = function() {
    return '$group in $select.groups track by $group.name';
  };

}]);


src/uiSelectChoicesDirective.js
uis.directive('uiSelectChoices',
  ['uiSelectConfig', 'uisRepeatParser', 'uiSelectMinErr', '$compile', '$window',
  function(uiSelectConfig, RepeatParser, uiSelectMinErr, $compile, $window) {

  return {
    restrict: 'EA',
    require: '^uiSelect',
    replace: true,
    transclude: true,
    templateUrl: function(tElement) {
      // Needed so the uiSelect can detect the transcluded content
      tElement.addClass('ui-select-choices');

      // Gets theme attribute from parent (ui-select)
      var theme = tElement.parent().attr('theme') || uiSelectConfig.theme;
      return theme + '/choices.tpl.html';
    },

    compile: function(tElement, tAttrs) {

      if (!tAttrs.repeat) throw uiSelectMinErr('repeat', "Expected 'repeat' expression.");

      // var repeat = RepeatParser.parse(attrs.repeat);
      var groupByExp = tAttrs.groupBy;
      var groupFilterExp = tAttrs.groupFilter;

      if (groupByExp) {
        var groups = tElement.querySelectorAll('.ui-select-choices-group');
        if (groups.length !== 1) throw uiSelectMinErr('rows', "Expected 1 .ui-select-choices-group but got '{0}'.", groups.length);
        groups.attr('ng-repeat', RepeatParser.getGroupNgRepeatExpression());
      }

      var parserResult = RepeatParser.parse(tAttrs.repeat);

      var choices = tElement.querySelectorAll('.ui-select-choices-row');
      if (choices.length !== 1) {
        throw uiSelectMinErr('rows', "Expected 1 .ui-select-choices-row but got '{0}'.", choices.length);
      }

      choices.attr('ng-repeat', parserResult.repeatExpression(groupByExp))
             .attr('ng-if', '$select.open'); //Prevent unnecessary watches when dropdown is closed


      var rowsInner = tElement.querySelectorAll('.ui-select-choices-row-inner');
      if (rowsInner.length !== 1) {
        throw uiSelectMinErr('rows', "Expected 1 .ui-select-choices-row-inner but got '{0}'.", rowsInner.length);
      }
      rowsInner.attr('uis-transclude-append', ''); //Adding uisTranscludeAppend directive to row element after choices element has ngRepeat

      // If IE8 then need to target rowsInner to apply the ng-click attr as choices will not capture the event.
      var clickTarget = $window.document.addEventListener ? choices : rowsInner;
      clickTarget.attr('ng-click', '$select.select(' + parserResult.itemName + ',$select.skipFocusser,$event)');

      return function link(scope, element, attrs, $select) {


        $select.parseRepeatAttr(attrs.repeat, groupByExp, groupFilterExp); //Result ready at $select.parserResult
        $select.disableChoiceExpression = attrs.uiDisableChoice;
        $select.onHighlightCallback = attrs.onHighlight;
        $select.dropdownPosition = attrs.position ? attrs.position.toLowerCase() : uiSelectConfig.dropdownPosition;

        scope.$watch('$select.search', function(newValue) {
          if(newValue && !$select.open && $select.multiple) $select.activate(false, true);
          $select.activeIndex = $select.tagging.isActivated || !$select.firstItemActive ? (newValue === '' ? -1 : 0) : 0;
          if (!attrs.minimumInputLength || $select.search.length >= attrs.minimumInputLength) {
            $select.refresh(attrs.refresh);
          } else {
            $select.items = [];
          }
        });

        attrs.$observe('refreshDelay', function() {
          // $eval() is needed otherwise we get a string instead of a number
          var refreshDelay = scope.$eval(attrs.refreshDelay);
          $select.refreshDelay = refreshDelay !== undefined ? refreshDelay : uiSelectConfig.refreshDelay;
        });

        attrs.$observe('firstItemActive', function() {
          // $eval() is needed otherwise we get a string instead of a boolean
          var firstItemActive = scope.$eval(attrs.firstItemActive);
          $select.firstItemActive = firstItemActive !== undefined ? firstItemActive : uiSelectConfig.firstItemActive;
        });

        scope.$watch('$select.open', function(open) {
          if (open) {
            tElement.attr('role', 'listbox');
            $select.refresh(attrs.refresh);
          } else {
            element.removeAttr('role');
          }
        });
      };
    }
  };
}]);


src/common.js
var KEY = {
    TAB: 9,
    ENTER: 13,
    ESC: 27,
    SPACE: 32,
    LEFT: 37,
    UP: 38,
    RIGHT: 39,
    DOWN: 40,
    SHIFT: 16,
    CTRL: 17,
    ALT: 18,
    PAGE_UP: 33,
    PAGE_DOWN: 34,
    HOME: 36,
    END: 35,
    BACKSPACE: 8,
    DELETE: 46,
    COMMAND: 91,

    MAP: { 91 : "COMMAND", 8 : "BACKSPACE" , 9 : "TAB" , 13 : "ENTER" , 16 : "SHIFT" , 17 : "CTRL" , 18 : "ALT" , 19 : "PAUSEBREAK" , 20 : "CAPSLOCK" , 27 : "ESC" , 32 : "SPACE" , 33 : "PAGE_UP", 34 : "PAGE_DOWN" , 35 : "END" , 36 : "HOME" , 37 : "LEFT" , 38 : "UP" , 39 : "RIGHT" , 40 : "DOWN" , 43 : "+" , 44 : "PRINTSCREEN" , 45 : "INSERT" , 46 : "DELETE", 48 : "0" , 49 : "1" , 50 : "2" , 51 : "3" , 52 : "4" , 53 : "5" , 54 : "6" , 55 : "7" , 56 : "8" , 57 : "9" , 59 : ";", 61 : "=" , 65 : "A" , 66 : "B" , 67 : "C" , 68 : "D" , 69 : "E" , 70 : "F" , 71 : "G" , 72 : "H" , 73 : "I" , 74 : "J" , 75 : "K" , 76 : "L", 77 : "M" , 78 : "N" , 79 : "O" , 80 : "P" , 81 : "Q" , 82 : "R" , 83 : "S" , 84 : "T" , 85 : "U" , 86 : "V" , 87 : "W" , 88 : "X" , 89 : "Y" , 90 : "Z", 96 : "0" , 97 : "1" , 98 : "2" , 99 : "3" , 100 : "4" , 101 : "5" , 102 : "6" , 103 : "7" , 104 : "8" , 105 : "9", 106 : "*" , 107 : "+" , 109 : "-" , 110 : "." , 111 : "/", 112 : "F1" , 113 : "F2" , 114 : "F3" , 115 : "F4" , 116 : "F5" , 117 : "F6" , 118 : "F7" , 119 : "F8" , 120 : "F9" , 121 : "F10" , 122 : "F11" , 123 : "F12", 144 : "NUMLOCK" , 145 : "SCROLLLOCK" , 186 : ";" , 187 : "=" , 188 : "," , 189 : "-" , 190 : "." , 191 : "/" , 192 : "`" , 219 : "[" , 220 : "\\" , 221 : "]" , 222 : "'"
    },

    isControl: function (e) {
        var k = e.which;
        switch (k) {
        case KEY.COMMAND:
        case KEY.SHIFT:
        case KEY.CTRL:
        case KEY.ALT:
            return true;
        }

        if (e.metaKey || e.ctrlKey || e.altKey) return true;

        return false;
    },
    isFunctionKey: function (k) {
        k = k.which ? k.which : k;
        return k >= 112 && k <= 123;
    },
    isVerticalMovement: function (k){
      return ~[KEY.UP, KEY.DOWN].indexOf(k);
    },
    isHorizontalMovement: function (k){
      return ~[KEY.LEFT,KEY.RIGHT,KEY.BACKSPACE,KEY.DELETE].indexOf(k);
    },
    toSeparator: function (k) {
      var sep = {ENTER:"\n",TAB:"\t",SPACE:" "}[k];
      if (sep) return sep;
      // return undefined for special keys other than enter, tab or space.
      // no way to use them to cut strings.
      return KEY[k] ? undefined : k;
    }
  };

/**
 * Add querySelectorAll() to jqLite.
 *
 * jqLite find() is limited to lookups by tag name.
 * TODO This will change with future versions of AngularJS, to be removed when this happens
 *
 * See jqLite.find - why not use querySelectorAll? https://github.com/angular/angular.js/issues/3586
 * See feat(jqLite): use querySelectorAll instead of getElementsByTagName in jqLite.find https://github.com/angular/angular.js/pull/3598
 */
if (angular.element.prototype.querySelectorAll === undefined) {
  angular.element.prototype.querySelectorAll = function(selector) {
    return angular.element(this[0].querySelectorAll(selector));
  };
}

/**
 * Add closest() to jqLite.
 */
if (angular.element.prototype.closest === undefined) {
  angular.element.prototype.closest = function( selector) {
    var elem = this[0];
    var matchesSelector = elem.matches || elem.webkitMatchesSelector || elem.mozMatchesSelector || elem.msMatchesSelector;

    while (elem) {
      if (matchesSelector.bind(elem)(selector)) {
        return elem;
      } else {
        elem = elem.parentElement;
      }
    }
    return false;
  };
}

var latestId = 0;

var uis = angular.module('ui.select', [])

.constant('uiSelectConfig', {
  theme: 'bootstrap',
  searchEnabled: true,
  sortable: false,
  placeholder: '', // Empty by default, like HTML tag <select>
  refreshDelay: 1000, // In milliseconds
  closeOnSelect: true,
  skipFocusser: false,
  dropdownPosition: 'auto',
  removeSelected: true,
  resetSearchInput: true,
  generateId: function() {
    return latestId++;
  },
  appendToBody: false,
  spinnerEnabled: false,
  spinnerClass: 'glyphicon-refresh ui-select-spin',
  backspaceReset: true,
  firstItemActive: true
})

// See Rename minErr and make it accessible from outside https://github.com/angular/angular.js/issues/6913
.service('uiSelectMinErr', function() {
  var minErr = angular.$$minErr('ui.select');
  return function() {
    var error = minErr.apply(this, arguments);
    var message = error.message.replace(new RegExp('\nhttp://errors.angularjs.org/.*'), '');
    return new Error(message);
  };
})

// Recreates old behavior of ng-transclude. Used internally.
.directive('uisTranscludeAppend', function () {
  return {
    link: function (scope, element, attrs, ctrl, transclude) {
        transclude(scope, function (clone) {
          element.append(clone);
        });
      }
    };
})

/**
 * Highlights text that matches $select.search.
 *
 * Taken from AngularUI Bootstrap Typeahead
 * See https://github.com/angular-ui/bootstrap/blob/0.10.0/src/typeahead/typeahead.js#L340
 */
.filter('highlight', function() {
  function escapeRegexp(queryToEscape) {
    return ('' + queryToEscape).replace(/([.?*+^$[\]\\(){}|-])/g, '\\$1');
  }

  return function(matchItem, query) {
    return query && matchItem ? ('' + matchItem).replace(new RegExp(escapeRegexp(query), 'gi'), '<span class="ui-select-highlight">$&</span>') : matchItem;
  };
})

/**
 * A read-only equivalent of jQuery's offset function: http://api.jquery.com/offset/
 *
 * Taken from AngularUI Bootstrap Position:
 * See https://github.com/angular-ui/bootstrap/blob/master/src/position/position.js#L70
 */
.factory('uisOffset',
  ['$document', '$window',
  function ($document, $window) {

  return function(element) {
    var boundingClientRect = element[0].getBoundingClientRect();
    return {
      width: boundingClientRect.width || element.prop('offsetWidth'),
      height: boundingClientRect.height || element.prop('offsetHeight'),
      top: boundingClientRect.top + ($window.pageYOffset || $document[0].documentElement.scrollTop),
      left: boundingClientRect.left + ($window.pageXOffset || $document[0].documentElement.scrollLeft)
    };
  };
}]);
